UI: Rework volume-meters, adding more information

takev opened the pull request:
The following features have been added to the audio-meters:
 * Stereo PPM-level meter, with 20 dB/1.7s decay rate.
 * Stereo VU-level meter, with 300 ms integration time.
 * Stereo Peak-hold meter, with 20 second sustain.
 * Input level peak level color-dot.
 * Minor-ticks for each dB.
 * Major-ticks for every 5 dB.
 * Meter is divided in sections at -20 dB and -9 dB.

The ballistic parameters chosen here where taken from:
 * https://en.wikipedia.org/wiki/Peak_programme_meter (SMPTE RP.0155)
 * https://en.wikipedia.org/wiki/VU_meter

In the rework I have removed any ballistic calculations from
libobs/obs-audio-controls.c making the calculations here a lot more
simple doing only MAX and RMS calculations for only the samples in
the current update. The actual ballistics are now done by just
the UI/volume-control.cpp because ballistics need to be updated
based on the repaint-rate of the user-interface.

The dB to pixel conversion has been moved from
libobs/obs-audio-controls.c to UI/volume-control.cpp as well to reduce
coupling between these two objects, especially when implementing the
major- and minor-ticks and the sections.

All colors and ballistic parameters are adjustable via QT style sheets.
There are slight differences in colors for each of the themes.

pkviet commented:
very nice; how hard would it be to generalize that for surround sound ?

takev commented:
It would not be super hard. libobs/obs-audio-controls already export the peak and magnitude for every plane. I was just thinking that maybe a boolean should be passed for each plane to know if it is enabled, which I think you can find out if the audio data for the plane is NULL or not.

The design of UI/volume-control.cpp also has a seperate function for drawing the meter. Together with the boolean it could draw 0 or more of them. I am not sure how QT will react by resize events generated by the paintEvent function though, because you would need to reserve more vertical space.

Fenrirthviti commented:
Would it be possible to create a separate, dockable item in the main UI that contains a full volume meter readout? Most users of OBS aren't going to care about this level of fidelity, and at worst it might confuse people. Having an optional, dockable window (i.e. some kind of audio control panel with this information) might be better from a UX standpoint? For stereo, it seems reasonable to have in the main OBS mixer, but for anything else it would get pretty messy pretty fast without a better way of displaying the information.

takev commented:
Would you like me to do the suround sound work?

Fenrirthviti, with a slight modification of how it shows, it only requires 4 pixels in height per audio-plane.

I think it would be reasonable to show all channels of a source from a UX standpoint, because it shows the user how many planes are actually being used by OBS instead of hiding this.

I haven't looked to deep into how audio is routed in OBS, can you already in the advanced options select which plane is used by an audio source?

dodgepong commented:
Assuming that [this](https://obsproject.com/forum/attachments/obs-meters-png.32847/) is a screenshot of the new bars, I don't believe it to be that confusing. Users might not know what the ticks and dot mean, but ignoring those, they would certainly know that higher is louder and red is bad, and that for a stereo meter, one bar is for L and one is for R. Perhaps this is a case where tooltips might be useful, just in case. Personally, I really like this level of information, especially the ticks and peak information.

takev commented:
Yes dodgepong, that is almost how it looks, I made the ticks look slightly better now.
I was thinking of having the ticks above the volume meters and below the volume meters. Then have all volume meters in the center with a pixel spacing between each.

pkviet commented:
only speaking for myself, but meters for surround would be great of course.
you can also go to irc #obs-dev for futher discussion since this is drawing the attention of several devs

takev commented:
Okay. I will update this pull request to:
 * Draw a meter for all enabled audio channels of audio source.
 * Draw a dot (probably a square) in front of each channel for the input-peak
 * Draw the ticks above and below the set of meters.
 * Draw all the ticks with a single drawLines() call for performance.

SuslikV commented:
I don't like it. There is no scale (values on the bar). No one of the wiki examples uses empty meters. It is not a garland on a Christmas tree.

WizardCM commented:
I'm curious. Currently OBS will mix down desktop audio that's surround (5.1 in my case) into Stereo for a Twitch stream _really_ well. How would this new (really cool-looking) volume meter treat that?

I also agree that having a secondary, larger mixer would be really useful with a design like this - one that can be popped out and displayed separately.

takev commented:
SuslikV, scale values do increase the height of the meters, but it may be worth it. However each tick is a single dB and full scale is 0dB.

takev commented:
Okay, I have updated this pull request. It now handles suround sound. It also adds numbers below the meters. Drawing of ticks and numbers are cached for performance.

Below is an example (except in reality the ticks are now even shorter).

![obs-meters](https://user-images.githubusercontent.com/303028/34469456-5dc395b0-ef1f-11e7-943f-16d86daa84f2.png)

pkviet commented:
wow, wonderful

takev commented:
Sadly, I can not really check how this patch looks on OS X and Windows. I hope it looks just as well.

ntoff commented:
Hell yeah! Adjusting the gate settings is a royal pain in the neck since OBS doesn't put numbers anywhere. How are you meant to set your input gate to a certain db level, when not one single location in OBS shows any kind of db number? I would love for this to become a part of OBS (even if it is some sort of optional extra).

reboot commented:
I tested the update and the meter bahviour seems a bit strange. If source is completely silence when I start OBS it doesn't seem to draw any meters for the source. It only draws the ticks. While audio is reveived it is ok, but if I stop the audio source again, then the meters will freeze like I would still receive the same audio levels I has when I stopped the audio. It does not go down to 0 again.

You can see the problem here: https://youtu.be/cnhiX2A_Wys

I'm using pulseaudio on Linux.

takev commented:
Interesting. I think I know a bit where the problem comes from. What kind of audio source are you using, maybe I can duplicate the issue and then fix it.

I think at the start the source is not only silent, but it is not sending audio at all. So the number of channels never get reported. I am not sure what is happening at the end, but I guess the sound is no longer being send.

reboot commented:
It is a null source that only receives audio from programs that should be streamed. Usually no source is sending audio to it. It only got audio while the music was playing.

takev commented:
reboot, I tried reproducing it, but I was not successful. However I think I know what is wrong and I will attempt a solution tomorrow.

 * A source that has no live audio stream yet, will have a single meter, so that it won't look weird.
 * Once a source has a live audio stream it will work like normal.
 * If there hasn't been a live audio stream for 1 second the number of meters will remain normal, but the value will be set to "no sound". So that it won't look like the meter is stuck.

reboot commented:
Can't you determine the number of planes/channels before a source receives audio?

takev commented:
The planes that are enabled are actually fixed for all source based on the global "Channels" setting in OBS.

The number of channels can variate at run time, while a source already exists. I could add a callback mechanism to react of of changes to the configuration of a source (which could also change without user interaction), but that seems extreme overkill.

takev commented:
I hope I fixed the issue that reboot had.
* Now checks how many channels there are on each redraw.
* The number is channels are: The number of channels of the source, clamped between 1 and global OBS channel configuration.
* The meters monitor if the audio source becomes idle. When it does it clears the meters and removes the little input-peak squares so that you can visually determine the difference between an idle audio stream and quiet audio.
* Updated the Rachni theme with Fenrir's new color scheme.

takev commented:
Hopefully the last update on this pull-request:
 * Double the speed of the peak meter decay rate to more closely match the speed of the original meters.

derrod commented:
Screenshot of the current version on Windows with default window dimensions:
![Current version](https://vgy.me/Z62240.png)

takev commented:
Thank you for the screenshot derrod. It looks correct.

ntoff commented:
do you have to compile OBS yourself to get it on windows? I'd love to try this out but don't have anything set up to compile it

derrod commented:
Right now: yes. Unless someone is willing to share a build or you clone it into your own repo and let AppVeyor compile it.

takev commented:
Modified pull request to fit within 80 character line limit.

WizardCM commented:
A wiki page has been created to help explain this system to users (once it's released, of course). I would still like to add graphs and images to the technical details section, but in the meantime would love feedback.

https://github.com/jp9000/obs-studio/wiki/Understanding-the-Mixer

takev commented:
I love the wiki page. Both easy to understand and a lot of detail when wanting more information.

jp9000 commented:
Top notch pull request.  A slight API change, but worth changing the API for.
